### PR TITLE
SCP-3493: change validation-full&decode to use the Ledger.Api

### PR DESCRIPTION
**Pull request by bezirg**
Changes the validation-full&decode from handcrafted code to Ledger.Api module calls.
I would also like to change the normal validation module to use the `Scripts.evaluateScript` but then it would incur a performance penaly since that function would have to account for (fakename pass + scope checking).

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by bezirg** [comment]
This needs polishing, but I wanted to show you the bench diff of this pr.

**Comment by kwxm** [comment]
@bezirg @michaelpj See https://github.com/input-output-hk/plutus/pull/4395 and SCP-2581.

**Comment by kwxm** [comment]
Thanks for doing this @bezirg: it's provided some useful information.  The difference between the two column shows how much time `applyCostModelParams` takes, about 4.1-4.5 ms.

File: plutus-ledger-api/src/Plutus/V1/Ledger/Scripts.hs [hunk 1]
```diff
@@ -57,6 +57,7 @@ module Plutus.V1.Ledger.Scripts(
     -- * Example scripts
     unitRedeemer,
     unitDatum,
+    SerialiseViaFlat (..)
```
Comment (michaelpj) [inline plutus-ledger-api/src/Plutus/V1/Ledger/Scripts.hs:60]: I don't think you should need this.

File: plutus-benchmark/validation/BenchFull.hs [hunk 2]
```diff
@@ -15,8 +25,11 @@ the whole time taken from script deserialization to script execution result.
 main :: IO ()
 main = benchWith mkFullBM
   where
-    mkFullBM file scriptBS = whnf (unsafeEvaluateCekNoEmit'
-                                  . throughCheckScope
-                                  . toNamedDeBruijnTerm
-                                  . unsafeUnflat file
-                                  ) scriptBS
+    mkFullBM _file flatBS =
+        let cbor = toShort $ toStrict $ Serialise.serialise $
```
Comment (michaelpj) [inline plutus-benchmark/validation/BenchFull.hs:29]: worth a small comment that we're just adjusting the serialization format here.

File: plutus-benchmark/validation/BenchFull.hs [hunk 3]
```diff
@@ -15,8 +25,11 @@ the whole time taken from script deserialization to script execution result.
...
+                       case unflat @(UPLC.Program UPLC.DeBruijn UPLC.DefaultUni UPLC.DefaultFun ()) flatBS of
+                           Left _e -> error "mpla"
```
Comment (michaelpj) [inline plutus-benchmark/validation/BenchFull.hs:31]: put something comprehensible here

**Review by michaelpj — APPROVED** [review]
Generally looks good, let's merge this.

**Comment by bezirg** [comment]
> Generally looks good, let's merge this.
@michaelpj Can we delay the merge for a bit? I was thinking to first to bring in https://github.com/input-output-hk/plutus/pull/4412  , otherwise it will dilute its performance uplift result

**Comment by michaelpj** [comment]
Sure, although I think we've seen the performance figures on that! As you lke.

**Comment by bezirg** [comment]
I think the decoding is faster because I am decoding via Script and not via ScriptForExecution

**Comment by bezirg** [comment]
> I think the decoding is faster because I am decoding via Script and not via ScriptForExecution

Actually that is wrong. It probably is that before we were doing asking an `nf` of the AST, but now which just ask if it can just deserialize correctly? :shrug:

**Comment by bezirg** [comment]
the bad validation-full results are expected. LGTM?

**Comment by michaelpj** [comment]
Hang on, I just want to try and understand what's going on with the decode benchmark.

**Comment by bezirg** [comment]
> Hang on, I just want to try and understand what's going on with the decode benchmark.

can it be that before we were using a lazy bytestring for flat deserialization whereas now we use a strict short bytestring?

File: plutus-core/plutus-core/test/CostModelInterface/Spec.hs [hunk 4]
```diff
@@ -91,6 +91,12 @@ testSelfUpdate model = do
   updated <- applyParams model params
   updated @?= model
 
+-- Update a model with its no parameters and check that we get the same model back
+testUpdateEmpty :: CekCostModel -> IO ()
```
Comment (michaelpj) [inline plutus-core/plutus-core/test/CostModelInterface/Spec.hs:95]: This seems unrelated? Fine, but maybe part of some other branch?

**Review by michaelpj — APPROVED** [review]
Eh, I think it's fine.

**Pull request merged**